File: NewsDaily2/app/src/main/java/com/andruid/magic/newsdaily/adapter/NewsAdapter.java
package com.andruid.magic.newsdaily.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.paging.PagedList;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;

import com.andruid.magic.newsdaily.databinding.LayoutNewsBinding;
import com.andruid.magic.newsdaily.viewholder.NewsViewHolder;
import com.andruid.magic.newsloader.model.News;

import java.util.ArrayList;
import java.util.List;

public class NewsAdapter extends PagedListAdapter<News, NewsViewHolder> {
    private static final DiffUtil.ItemCallback<News> DIFF_CALLBACK = new DiffUtil.ItemCallback<News>() {
        @Override
        public boolean areItemsTheSame(@NonNull News oldItem, @NonNull News newItem) {
            return oldItem.getTitle().compareTo(newItem.getTitle())==0;
        }

        @Override
        public boolean areContentsTheSame(@NonNull News oldItem, @NonNull News newItem) {
            return oldItem.equals(newItem);
        }
    };

    public NewsAdapter(){
        super(DIFF_CALLBACK);
    }

    @NonNull
    @Override
    public NewsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        LayoutNewsBinding binding = LayoutNewsBinding.inflate(inflater, parent, false);
        return new NewsViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull NewsViewHolder holder, int position) {
        News news = getItem(position);
        holder.bind(news);
    }

    public List<News> getNewsList(){
        PagedList<News> newsPagedList = getCurrentList();
        if(newsPagedList != null)
            return new ArrayList<>(newsPagedList.snapshot());
        return new ArrayList<>();
    }
}

File: NewsDaily/app/src/main/java/com/death14stroke/newsdaily/ui/adapter/NewsAdapter.kt
package com.death14stroke.newsdaily.ui.adapter

import android.view.ViewGroup
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import com.death14stroke.newsdaily.data.model.OpenNewsListener
import com.death14stroke.newsdaily.data.model.ShareNewsListener
import com.death14stroke.newsdaily.data.model.ViewImageListener
import com.death14stroke.newsdaily.ui.viewholder.NewsViewHolder
import com.death14stroke.newsloader.data.model.News

private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<News>() {
    override fun areContentsTheSame(oldItem: News, newItem: News) =
        oldItem.url == newItem.url

    override fun areItemsTheSame(oldItem: News, newItem: News) =
        oldItem == newItem
}

class NewsAdapter(private val viewImageListener: ViewImageListener, private val openNewsListener: OpenNewsListener, private val shareNewsListener: ShareNewsListener) :
    PagingDataAdapter<News, NewsViewHolder>(DIFF_CALLBACK) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        NewsViewHolder.from(parent)

    override fun onBindViewHolder(holder: NewsViewHolder, position: Int) {
        getItem(position)?.let { news ->
            holder.bind(news, viewImageListener, openNewsListener, shareNewsListener)
        }
    }
}